src/control/ros/RosTrajectoryExecutor.cpp

```diff
@@ -0,0 +1,249 @@
+#include <aikido/control/ros/Conversions.hpp>
+#include <aikido/control/ros/RosTrajectoryExecutor.hpp>
+#include <aikido/control/ros/RosTrajectoryExecutionException.hpp>
+#include <aikido/statespace/dart/MetaSkeletonStateSpace.hpp>
+#include <aikido/statespace/dart/RnJoint.hpp>
+#include <aikido/statespace/dart/SO2Joint.hpp>
+#include <aikido/util/StepSequence.hpp>
+
+namespace aikido {
+namespace control {
+namespace ros {
+namespace {
+
+//=============================================================================
+std::string getFollowJointTrajectoryErrorMessage(int32_t errorCode)
+{
+  using Result = control_msgs::FollowJointTrajectoryResult;
+
+  switch (errorCode)
+  {
+    case Result::SUCCESSFUL:
+      return "successful";
+
+    case Result::INVALID_GOAL:
+      return "invalid goal";
+
+    case Result::INVALID_JOINTS:
+      return "invalid joints";
+
+    case Result::OLD_HEADER_TIMESTAMP:
+      return "old header timestamp";
+
+    case Result::PATH_TOLERANCE_VIOLATED:
+      return "path tolerance violated";
+
+    case Result::GOAL_TOLERANCE_VIOLATED:
+      return "goal tolerance violated";
+
+    default:
+      std::stringstream errorString;
+      errorString << "unknown (" << errorCode << ")";
+      return errorString.str();
+  }
+}
+
+} // namespace
+
+//=============================================================================
+RosTrajectoryExecutor::RosTrajectoryExecutor(
+      statespace::dart::MetaSkeletonStateSpacePtr space,
+      ::ros::NodeHandle node,
+      const std::string& serverName,
+      double timestep,
+      double goalTimeTolerance,
+      const std::map<std::string, size_t>& jointIndexMap,
+      std::chrono::milliseconds connectionTimeout,
+      std::chrono::milliseconds connectionPollingPeriod)
+  : mSpace{std::move(space)}
+  , mNode{std::move(node)}
+  , mCallbackQueue{}
+  , mClient{mNode, serverName, &mCallbackQueue}
+  , mTimestep{timestep}
+  , mGoalTimeTolerance{goalTimeTolerance}
+  , mConnectionTimeout{connectionTimeout}
+  , mConnectionPollingPeriod{connectionPollingPeriod}
+  , mInProgress{false}
+  , mJointIndexMap(jointIndexMap)
+{
+  if (!mSpace)
+    throw std::invalid_argument("Space is null.");
+
+  if (mTimestep <= 0)
+    throw std::invalid_argument("Timestep must be positive.");
+
+  if (mGoalTimeTolerance <= 0)
+    throw std::invalid_argument("Goal time tolerance must be positive.");
+}
+
+//=============================================================================
+RosTrajectoryExecutor::~RosTrajectoryExecutor()
+{
+  // Do nothing.
+  // TODO: Should we wait for the current trajectory to finish executing?
+}
+
+//=============================================================================
+std::future<void> RosTrajectoryExecutor::execute(
+  trajectory::TrajectoryPtr traj)
+{
+  static const ::ros::Time invalidTime;
```

Comment:
Passing invalid time is intentional? I think we could simply pass the current time using `ros::Time::now()`.